src/lib.rs

```diff
@@ -277,73 +365,91 @@ impl<T: Sync + Send + Clone> Clone for SeccSender<T> {
+        }
+
+        // Lock the channel to operate on it.
+        let (ref mutex, _, has_data) = &*self.core.lock;
+        let mut pointers = mutex.lock().unwrap();
 
         // Get a pointer to the current pool_head and see if we have space to send.
-        let pool_head_ptr = &self.core.nodes[send_ptrs.pool_head];
+        let pool_head_ptr = &self.core.nodes[pointers.pool_head];
         let next_pool_head = pool_head_ptr.next.load(Ordering::SeqCst);
         if NIL == next_pool_head {
+            if let Some(w) = waker {
+                if let Some(ref mut wakers) = pointers.send_wakers {
+                    wakers.push(w);
+                } else {
+                    pointers.send_wakers = Some(vec![w]);
+                }
+            }
             Err(SeccErrors::Full(message))
         } else {
             // We get the queue tail because the node from the pool will move here.
-            let queue_tail_ptr = &self.core.nodes[send_ptrs.queue_tail];
+            let queue_tail_ptr = &self.core.nodes[pointers.queue_tail];
 
             // Add the message to the node, transferring ownership.
             unsafe {
                 *queue_tail_ptr.cell.get() = Some(message);
             }
 
             // Update the pointers in the mutex.
-            let old_pool_head = send_ptrs.pool_head;
-            send_ptrs.queue_tail = send_ptrs.pool_head;
-            send_ptrs.pool_head = next_pool_head;
-
-            // Adjust the channel metrics.
-            self.core.sent.fetch_add(1, Ordering::SeqCst);
-            self.core.receivable.fetch_add(1, Ordering::SeqCst);
-            self.core.pending.fetch_add(1, Ordering::SeqCst);
+            let old_pool_head = pointers.pool_head;
+            pointers.queue_tail = pointers.pool_head;
+            pointers.pool_head = next_pool_head;
 
             // The now filled node will get moved to the queue.
             pool_head_ptr.next.store(NIL, Ordering::SeqCst);
-
-            // We MUST set this LAST or we will get into a race with the receiver that would
-            // think this node is ready for receiving when it isn't until just now.
             queue_tail_ptr.next.store(old_pool_head, Ordering::SeqCst);
 
+            // Take the receive wakers if any for iteration outside of the guard.
+            let wakers_opt = pointers.receive_wakers.take();
+
             // Notify anyone that was waiting on the Condvar and we are done.
-            condvar.notify_all();
+            has_data.notify_all();
+
+            // Adjust receivable count.
+            self.core.receivable.fetch_add(1, Ordering::SeqCst);
+
+            // End of critical block so drop the lock.
+            drop(pointers);
+
+            // Adjust the channel metrics.
+            self.core.sent.fetch_add(1, Ordering::SeqCst);
+            self.core.pending.fetch_add(1, Ordering::SeqCst);
+
+            // Call any potential wakers.
+            if let Some(mut wakers) = wakers_opt {
```

Comment:
I would like to ask someone who's much more comfortable with Futures to review this consumption of Wakers. It looks alright, but I'm still a little unsure.

Reply:
There are alternatives as pointedo out on discord but no compellign reason to change this.